File: src/0572-subtree-of-another-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
    // we've reached the end of the tree without finding a match, so its false
    if (!root) {
        return false;
    }

    // an empty tree is considered a subtree of any tree, so we return true.
    if (!subRoot) {
        return true;
    }

    if (isSameTree(root, subRoot)) {
        return true;
    }

    // checks if either of the left/right of the main tree contains the subtree
    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    if (!p && !q) {
        return true;
    }

    if (p && q && p.val === q.val) {
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }

    return false;
}

File: src/0572-subtree-of-another-tree/solution.py
from collections import deque

class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        
        # BFS queue initialized with the root of the main tree
        queue = deque([root])
        
        while queue:
            node = queue.popleft()
            
            # Check if the current node's subtree matches subRoot
            if self.isSameTree(node, subRoot):
                return True
            
            # Add the left and right children to the queue to explore next
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        # if we reached here means we ran out of nodes to check after searching through all nodes
        return False
    
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False
        
        # Check both left and right subtrees recursively
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
